dnf/cli/commands/history.py

```diff
@@ -0,0 +1,506 @@
+                )
+
+            query = dnf.subject.Subject(nevra).get_best_solution(
+                self.base.sack,
+                with_nevra=True,
+                with_provides=False,
+                with_filenames=False,
+                forms=[hawkey.FORM_NEVRA]
+            )["query"]
+
+            if not query:
+                raise TransactionError(_('Package "{nevra}" was not found.').format(nevra=nevra))
+
+            if action in ("Install", "Upgrade", "Downgrade"):
+                sltr = dnf.selector.Selector(self.base.sack).set(pkg=query)
+                self.base.goal.install(select=sltr, optional=False)
+            elif action == "Reinstall":
+                sltr = dnf.selector.Selector(self.base.sack).set(pkg=query.available())
+                self.base.goal.install(select=sltr, optional=False)
+            elif action == "Removed" or action == "Obsoleted":
+                query = query.installed()
+
+                if not query:
+                    raise TransactionError(
+                        _('Cannot remove package "{nevra}", it is not installed.').format(nevra=nevra)
+                    )
+
+                self.base.goal.erase(query[0])
+        except KeyError as e:
+            raise TransactionError(
+                _('Missing object key "{key}" under package "{pkg}".').format(key=e.args[0], pkg=na_str)
+            )
+
+
+    def replay_transaction(self):
+        fn = self.opts.transaction_filename
+
+        # Override configuration options that affect how the transaction is resolved
+        self.base.conf.clean_requirements_on_remove = False
+        self.base.conf.install_weak_deps = False
+
+        with open(fn, "r") as f:
+            try:
+                self.replay_data = json.load(f)
+            except json.decoder.JSONDecodeError as e:
+                raise TransactionFileError(fn, str(e) + ".")
+
+        self._verify_toplevel_json()
+
+        rpms = self.replay_data.get("rpms", {})
+        if not isinstance(rpms, dict):
+            raise TransactionFileError(fn, _('Unexpected type of "rpms", object expected.'))
+
+        groups = self.replay_data.get("groups", {})
+        if not isinstance(groups, dict):
+            raise TransactionFileError(fn, _('Unexpected type of "groups", object expected.'))
+
+        environments = self.replay_data.get("environments", {})
+        if not isinstance(environments, dict):
+            raise TransactionFileError(fn, _('Unexpected type of "environments", object expected.'))
+
+        self._na_reason_cache = {}
+        errors = []
+
+        for na, pkg_data in rpms.items():
+            subj = hawkey.Subject(na)
+            nas = subj.get_nevra_possibilities(forms=[hawkey.FORM_NA])
+
+            if len(nas) != 1:
+                errors.append(TransactionError(_('Cannot parse name.arch for package "{na}".').format(na=na)))
+                continue
+
+            if isinstance(pkg_data, list):
```

Comment:
This part could be simplified when NA dict in JSON will be replaced by list of dicts.